src/main/scala/is/hail/variant/VariantDataset.scala

```diff
@@ -44,59 +50,93 @@ object VariantDataset {
     val isGenericGenotype = metadata.isGenericGenotype
     val isDosage = metadata.isDosage
 
-    if (isGenericGenotype)
-      fatal("Cannot read datasets with generic genotypes.")
+    if (!parquetGenotypes && !isGenericGenotype) {
+      return new VariantDataset(hc,
+        metadata,
+        MatrixRead(hc, dirname, metadata, dropSamples = skipGenotypes, dropVariants = skipVariants,
+          MatrixType(metadata)))
+    }
 
     val parquetFile = dirname + "/rdd.parquet"
 
-    val orderedRDD = if (skipVariants)
-      OrderedRDD.empty[Locus, Variant, (Annotation, Iterable[Genotype])](sc)
-    else {
-      val rdd = if (skipGenotypes)
-        sqlContext.readParquetSorted(parquetFile, Some(Array("variant", "annotations")))
-          .map(row => (row.getVariant(0),
-            (if (vaRequiresConversion) SparkAnnotationImpex.importAnnotation(row.get(1), vaSignature) else row.get(1),
-              Iterable.empty[Genotype])))
+    val partitioner: OrderedPartitioner[Locus, Variant] =
+      try {
+        val jv = hConf.readFile(dirname + "/partitioner.json.gz")(JsonMethods.parse(_))
+        jv.fromJSON[OrderedPartitioner[Locus, Variant]]
+      } catch {
+        case _: FileNotFoundException =>
+          fatal("missing partitioner.json.gz when loading VDS, create with HailContext.write_partitioning.")
+      }
+
+    val localValue =
+      if (skipGenotypes)
+        fileMetadata.localValue.dropSamples()
+      else
+        fileMetadata.localValue
+
+    if (isGenericGenotype) {
+      assert(!parquetGenotypes)
+
+      val orderedRDD = if (skipVariants)
+        OrderedRDD.empty[Locus, Variant, (Annotation, Iterable[Annotation])](sc)
       else {
-        val rdd = sqlContext.readParquetSorted(parquetFile)
-        if (parquetGenotypes)
-          rdd.map { row =>
-            val v = row.getVariant(0)
-            (v,
+        val rdd = if (skipGenotypes)
+          sqlContext.readParquetSorted(parquetFile, Some(Array("variant", "annotations")))
+            .map(row => (row.getVariant(0),
               (if (vaRequiresConversion) SparkAnnotationImpex.importAnnotation(row.get(1), vaSignature) else row.get(1),
-                row.getSeq[Row](2).lazyMap { rg =>
-                  new RowGenotype(rg): Genotype
-                }))
-          } else
+                Iterable.empty[Annotation])))
+        else {
+          val rdd = sqlContext.readParquetSorted(parquetFile)
           rdd.map { row =>
             val v = row.getVariant(0)
```

Comment:
This is the `isGenericGenotype` case, the `row.getGenotypeStream` method isn't used in it.